### ISSUE TITLE
Cant get authorized

### ISSUE THREAD

**Issue by pluckyhd** [ahaenggli/AzureAD-LDAP-wrapper#45]
Everything seems to be working except the ldap bind user I setup won't authorize. 

When setting up I can't browse/set the DN when trying to connect I get the error
{0} doesn't exist on the LDAP server. It may be removed by administrator. Please refresh the page and try again later

I think it pertains to this 
"LDAP search on the NAS must be possible without any authentication in order to be able to select the domain/baseDN at all"

but I am not sure how to enable it without authentication. I am sure I am missing something simple. 

If I manually type in there it accepts but then fails to authorize using the binduser I setup

**Comment by ahaenggli** [comment]
What DSM version are you running? Which env vars did you use for the wrapper?
Are there any errors in your docker log?

**Comment by pluckyhd** [comment]
Thanks for quick reply
DSM 7.1.1-42962 Update 4
Not sure which one your asking exactly  I have all the appid,tenantid,secret set according lets say my domain is user-fun.com
NODE_ENV = production
LDAP_DOMAIN = user-fun.com 
LDAP_BASEDN = dc=user-fun,dc=com
LDAP_BINDUSER=ldapuser|Test123   -> This is what I enter in ldap setup of NAS
GRAPH_IGNORE_MFA_ERRORS = true
DSM7 = true
LDAP_PORT = 13389

if you need anything else let me know

The error comes when I go through the ldap wizard (looks a bit different than your screenshots
I hit the ldap server fine but once putting in my username and password I can't select the BaseDN I have attached a screenshot the red section is where the error occurs.
![ldap_error](https://user-images.githubusercontent.com/5316103/222001026-daed61ca-ca9b-4d0e-a7a9-35742be7e9dc.jpg)

**Comment by pluckyhd** [comment]
Ok so found in the log upon starting


ERROR: 2023-03-01T03:19:45.792Z:  ldapwrapper.js ClientAuthError: network_error: Network request failed. Please check network trace to determine root cause. \| Fetch client threw: Error: HTTP status code 401 \| Attempted to reach: https://login.microsoftonline.com/{appidfromazure{/oauth2/v2.0/token | stdout
-- | --

I edited the log to hide the appid

</div><div class="x-grid3-row" role="row" style="margin: 0px; padding: 0px; cursor: default; border-width: 0px 0px 1px; border-style: solid; border-top-color: initial; border-right-color: initial; border-bottom-color: rgba(198, 212, 224, 0.4); border-left-color: initial; border-image: initial; width: 657px; background-color: transparent; color: rgb(65, 75, 85); font-size: 13px; height: 24px;"><table class="x-grid3-row-table" border="0" cellspacing="0" cellpadding="0" style="table-layout: fixed; width: 657px; height: 24px;"><tbody><tr role="row"><td class="x-grid3-col x-grid3-cell x-grid3-td-time x-grid3-cell-first " role="gridcell" tabindex="-1" id="ext-gen7561" aria-readonly="true" style="margin: 0px; padding: 0px; outline: none; line-height: 13px; vertical-align: top; height: 24px; color: rgb(65, 75, 85); font-family: Verdana, Arial, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; width: 79px;"><div class="x-grid3-cell-inner x-grid3-col-time x-unselectable" unselectable="on" style="margin: 0px; padding: 0px 8px; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; color: rgb(65, 75, 85); font-size: 13px; width: auto; line-height: 24px;">21:19:45</div></td><td class="x-grid3-col x-grid3-cell x-grid3-td-1 " role="gridcell" tabindex="-1" id="ext-gen7562" aria-readonly="true" style="margin: 0px; padding: 0px; outline: none; line-height: 13px; vertical-align: top; height: 24px; color: rgb(65, 75, 85); font-family: Verdana, Arial, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; width: 490px;"><div class="x-grid3-cell-inner x-grid3-col-1 x-unselectable" unselectable="on" style="margin: 0px; padding: 0px 8px; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; color: rgb(65, 75, 85); font-size: 13px; width: auto; line-height: 24px;"><br class="Apple-interchange-newline">
</body>
</html>

**Comment by ahaenggli** [comment]
"HTTP status code 401" means "Unauthorized". Are you using a proxy?
If not, please check the values of AZURE_APP_ID and AZURE_APP_SECRET. They do not match your AZURE_TENANTID. Maybe you copied too many spaces somewhere or something like that?

**Comment by pluckyhd** [comment]
Yay I have those correct I double checked them. I don't need to setup any user permissions on the azure side right? Just the read/delegate permissions listed for the app. Also I don't need a valid user on the docker side right (of azure that is it logs in im assuming by the secret)

**Comment by ahaenggli** [comment]
Did you follow [required settings in azuread](https://github.com/ahaenggli/AzureAD-LDAP-wrapper#required-settings-in-azuread)? Only those 3 permissions are needed...
Maybe the status for those permission ist "NOT granted" instead of "granted" in your portal?

I will add some more checks when starting the wrapper to easier isolate such problems in the future.

**Comment by pluckyhd** [comment]
Yes I  have that and its granted. I honestly have no idea I deleted the entire container and tried setting up from scratch and still get the same authorization error. If I try and access the link in the log I get an error because it expects a post. If you could tell me what you post I can verify that as I am capable of making a post request.

**Comment by ahaenggli** [comment]
Could you temporarily use the docker "dev" image instead of "latest"? In the dev tag are now some settings checked at startup. 
There should be some additional logs, which will maybe help us.

**Comment by pluckyhd** [comment]
Ok so ran dev

```

INFO: 2023-03-05T22:32:07.699Z:  graph_azuread.js system.loggerOptions 3 [Sun, 05 Mar 2023 22:32:07 GMT] : [899a63fe-5d05-441d-b6b6-6be80ac6a7f0] : @azure/msal-node@1.15.0 : Verbose - Client credential client created | stdout
-- | --
```

`
ERROR: 2023-03-05T22:32:07.978Z:  graph_azuread.js checkToken { | stdout
-
errorCode: 'invalid_client', | stdout
-- | --

errorMessage: "7000215 - [2023-03-05 22:32:01Z]: AADSTS7000215: Invalid client secret provided. Ensure the secret being sent in the request is the client secret value, not the client secret ID, for a secret added to app 'appid'.\r\n" + | stdout
-- | --
'
That is what I see ofcourse I blanked out the appid up there for security reasons




`

**Comment by ahaenggli** [comment]
This is great, thanks for trying! It looks like your value for `AZURE_APP_SECRET` is wrong (a common mistake would be to copy SecretId instead of the value) or the secret has expired. Can you create a new secret and use its value?

**Comment by pluckyhd** [comment]
Wow thanks can't believe I missed that but you were correct. As a fellow programmer I can't thank you enough for spending the time to add logging and resolve!
